consensus/blake3pow/poem.go

```diff
@@ -0,0 +1,125 @@
+package blake3pow
+
+import (
+	"math/big"
+
+	"github.com/dominant-strategies/go-quai/common"
+	"github.com/dominant-strategies/go-quai/core/types"
+	"github.com/dominant-strategies/go-quai/params"
+	"modernc.org/mathutil"
+)
+
+// CalcOrder returns the order of the block within the hierarchy of chains
+func (blake3pow *Blake3pow) CalcOrder(header *types.Header) (*big.Int, int, error) {
+	// Verify the seal and get the powHash for the given header
+	powHash, err := blake3pow.verifySeal(header)
+	if err != nil {
+		return big0, -1, err
+	}
+
+	// Get entropy reduction of this header
+	intrinsicS := blake3pow.IntrinsicLogS(powHash)
+
+	// This is the updated the threshold calculation based on the zone difficulty threshold
+	target := new(big.Int).Div(big2e256, header.Difficulty()).Bytes()
+	zoneThresholdS := blake3pow.IntrinsicLogS(common.BytesToHash(target))
+	timeFactorHierarchyDepthMultiple := new(big.Int).Mul(params.TimeFactor, big.NewInt(common.HierarchyDepth))
+
+	// Prime case
+	primeEntropyThreshold := new(big.Int).Mul(timeFactorHierarchyDepthMultiple, timeFactorHierarchyDepthMultiple)
+	primeEntropyThreshold = new(big.Int).Mul(primeEntropyThreshold, zoneThresholdS)
+	primeBlockThreshold := new(big.Int).Quo(primeEntropyThreshold, big.NewInt(2))
+	primeEntropyThreshold = new(big.Int).Sub(primeEntropyThreshold, primeBlockThreshold)
+
+	primeBlockEntropyThresholdAdder, _ := mathutil.BinaryLog(primeBlockThreshold, 8)
+	primeBlockEntropyThreshold := new(big.Int).Add(zoneThresholdS, big.NewInt(int64(primeBlockEntropyThresholdAdder)))
+
+	totalDeltaS := new(big.Int).Add(header.ParentDeltaS(common.REGION_CTX), header.ParentDeltaS(common.ZONE_CTX))
+	totalDeltaS.Add(totalDeltaS, intrinsicS)
+	if intrinsicS.Cmp(primeBlockEntropyThreshold) > 0 && totalDeltaS.Cmp(primeEntropyThreshold) > 0 {
+		return intrinsicS, common.PRIME_CTX, nil
+	}
+
+	// Region case
+	regionEntropyThreshold := new(big.Int).Mul(timeFactorHierarchyDepthMultiple, zoneThresholdS)
+	regionBlockThreshold := new(big.Int).Quo(regionEntropyThreshold, big.NewInt(2))
+	regionEntropyThreshold = new(big.Int).Sub(regionEntropyThreshold, regionBlockThreshold)
+
+	regionBlockEntropyThresholdAdder, _ := mathutil.BinaryLog(regionBlockThreshold, 8)
+	regionBlockEntropyThreshold := new(big.Int).Add(zoneThresholdS, big.NewInt(int64(regionBlockEntropyThresholdAdder)))
+
+	totalDeltaS = new(big.Int).Add(header.ParentDeltaS(common.ZONE_CTX), intrinsicS)
+	if intrinsicS.Cmp(regionBlockEntropyThreshold) > 0 && totalDeltaS.Cmp(regionEntropyThreshold) > 0 {
+		return intrinsicS, common.REGION_CTX, nil
+	}
+
+	return intrinsicS, common.ZONE_CTX, nil
```

Comment:
didn't we have an issue here with sealer saying a block is valid and we would reject here? I think we should use one reduced representation to decide if a block is zone/valid block

Reply:
That issue was actually because CalcIntrinsicS was using block hash instead of progpow hash.

If this were HLCR, I would agree with you that we should use one impl of the difficulty check; i.e. use `CalcOrder` in the sealer AND the verifier. Unfortunately, since POEM, CalcOrder does a lot more than just checking a difficulty threshold, and would cause an unnecessary performance hit to the miner. Not to mention the additional math is context dependent, but the miner does not have a context.

tl;dr: yes, there was an issue once, but its not an issue anymore

Reply:
Thats good, only difference would be 2 different methods, but this is good for now